### ISSUE TITLE
[Bug]: unsetFontFamily and unsetColor removes all styles in the selection

### ISSUE THREAD

**Issue by bluedot74** [ueberdosis/tiptap#4853]
### Which packages did you experience the bug in?

extension-text-style

### What Tiptap version are you using?

2.1.12

### What’s the bug you are facing?

The actual culprit is `removeEmptyTextStyle()` which is run on `unsetFontSize` and `unsetFontFamily` and presumably anything else that extends `textStyle`.

If you have HTML that looks like this:
`<p><span style="color: red">her</span>e <span style="font-family: Courier\ New">is</span> some text</p>`

And you run:
`editor.chain().selectAll().unsetColor().run();`

Both spans will be removed and you'll end up with this:
`<p>here is some text</p>`

The expectation is that only the span with the `color: red` would be removed, and the `font-family` span would be left alone.

I believe this is because [getMarkAttributes](https://github.com/ueberdosis/tiptap/blob/main/packages/core/src/helpers/getMarkAttributes.ts) 

Finds the first mark in the selection by name, and then returns it's attributes, which are then used to determine if *all* marks in the selection have attributes. 

So in my example above, `unsetColor()` will remove the color and leave the span's style empty: `<span style="">`. Then `getMarkAttributes` gathers all of the marks in the selection, and uses `marks.find()` to find the first one by name, which in this case will be `textStyle`. 
The first mark it finds with `marks.find()` is the newly empty `<span>` that had the `color`. It then returns that mark's attributes which will be an object with all falsey values.
`removeEmptyTextStyle` then sees that the mark is empty and calls `commands.unsetMark` which wipes out all of the `textStyle` marks, including the `<span style="font-family: Courier\ New">`

If the first span in the selection has an attribute non spans are removed.

### What browser are you using?

Chrome

### Code example

https://codesandbox.io/p/sandbox/silly-zhukovsky-t9t4nm?file=%2Fsrc%2FApp.js

### What did you expect to happen?

I expect `removeEmptyTextStyle()` to only remove spans with empty inline attributes

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by sjdemartini** [comment]
This is a duplicate of https://github.com/ueberdosis/tiptap/issues/4311

**Comment by www-chique** [comment]
The bug is still there after the [supposed fix](https://github.com/Senobraz/tiptap/commit/d735cf3c758d52ecb2273e45cc0c9727dbf6eccd).

@sjdemartini - This is a duplicate of #4311, but that thread is closed with a fix that has a flaw as mentioned in this thread.

**Comment by jakeorr** [comment]
First off, thanks for all the work on this library!

I'm seeing the same behaviour described in the issue with version 2.11.5. `unsetColor` is clearing other styles when run on text in an unordered list:

1. Enter text
2. Convert to an unordered list
3. Change font size of list item text
4. `unsetColor` clears the font size

Does anyone have info on when this might be addressed, or a potential workaround for now?

**Comment by jakeorr** [comment]
Here's what I've found digging into the code:

This is the html in my editor: `<ul><li><p><span style="font-size: 48px">asdf</span></p></li></ul>`

I have the "asdf" text highlighted and I'm triggered an `unsetColor` command on it.

1. When [iterating over the nodes](https://github.com/ueberdosis/tiptap/blob/4f498944b507d03da967c5d6462f72d02422fcb4/packages/extension-text-style/src/text-style/index.ts#L119) we see 4 nodes: 
  a. 'bulletList' - no marks, this results in a `removeMark` [call](https://github.com/ueberdosis/tiptap/blob/4f498944b507d03da967c5d6462f72d02422fcb4/packages/extension-text-style/src/text-style/index.ts#L134)
  b. 'listItem' - no marks, this results in a `removeMark` [call](https://github.com/ueberdosis/tiptap/blob/4f498944b507d03da967c5d6462f72d02422fcb4/packages/extension-text-style/src/text-style/index.ts#L134)
  c. 'paragraph' - skipped because of the `node.isTextblock` [check](https://github.com/ueberdosis/tiptap/blob/4f498944b507d03da967c5d6462f72d02422fcb4/packages/extension-text-style/src/text-style/index.ts#L122)
  d. 'text' - This is the one that has marks and it correctly does not [call](https://github.com/ueberdosis/tiptap/blob/4f498944b507d03da967c5d6462f72d02422fcb4/packages/extension-text-style/src/text-style/index.ts#L134) `removeMark`

When I add code that causes 'bulletList' and 'listItem' nodes to be skipped, the issue does not occur. Which makes sense because `removeMark` does not get called.

I think what's happening is that when `removeMark` is called on the parent nodes (bulletList and listItem), it is removing that mark type (textStyle) from all nodes in the range including children. So it is checking each node and correcting not calling `removeMark` for the node with attributes, but the parent nodes calling `removeMark` blows away the child mark anyway.

From the description of `removeMark`:

> Remove marks from inline nodes between from and to. When mark is a single mark, remove precisely that mark. When it is a mark type, remove all marks of that type. When it is null, remove all marks of any type.

Potential solutions:
1. Call `removeMark` in a different way so that it only removes marks on the current node. Is this already supported?
2. Block more than the 'paragraph' node type. This doesn't seem very elegant and I'm not sure if it would have other repercussions. Are there times when we do want to remove marks from bulletList and listItem nodes? This would also probably be a can of worms where you have to specify all kinds of node types.

Someone more familiar with the library might have a better suggestion.

@nperez0111 I'd love to hear if you have any thoughts on this, particularly if what I'm thinking in solution (1) is already supported.

**Comment by nicwegner** [comment]
@jakeorr You are right, I noticed the same thing. As you already said, the problem lies at this check:
```ts
tr.doc.nodesBetween(selection.from, selection.to, (node, pos) => {
// Check if it's a paragraph element, if so, skip this node as we apply
// the text style to inline text nodes only (span).
if (node.isTextblock) {
   return true;
}
```

`tr.doc.nodesBetween` starts at the document node, so everything inside the given range (even above the paragraph node like a table or list) gets invoked. `node.isTextblock` doesnt filter out the wrapping nodes like lists or tables since they still contain text. `removeMark` then removes the textStyle marks for the whole list/table.

Correct me if I am wrong but wouldn't the most simple solution be to change check to this?
```ts
if (!node.isText) {
   return true;
}
```
This way, only the Text nodes are the ones that are not skipped. This seems to work for me.

**Comment by jakeorr** [comment]
Hi @nicwegner, thanks for detailing what you've tried. That's the route that I went for our project. I changed that first clause to:

```javascript
if (node.isTextblock || node.type.name !== 'text') { 
    return true;
}
```

So that parent nodes do not clear textStyle marks on their children. Seems to be working for us.

**Comment by Jimbolino** [comment]
We hit this in production (inline font-weight per word inside a bullet list —
styling a second word silently erased the first) and can confirm the analysis by
@jakeorr and @nicwegner. Two additions that may help get this fixed:

**1. Minimal repro against the current implementation** (reproduced on
`@tiptap/extension-text-style` 2.27.2; the identical `removeEmptyTextStyle` code
is on v3 `main` in `packages/extension-text-style/src/text-style/index.ts`, so
this affects v3 as well):

```js
import { Editor } from '@tiptap/core'
import StarterKit from '@tiptap/starter-kit'
import TextStyle from '@tiptap/extension-text-style'
import Color from '@tiptap/extension-color'

const editor = new Editor({
    extensions: [StarterKit, TextStyle, Color],
    content:
        '<ul>' +
        '<li><p><span style="color: rgb(255, 0, 0)">alpha</span> beta</p></li>' +
        '<li><p>gamma delta</p></li>' +
        '</ul>',
})

// select the word "gamma" — in the SECOND list item — and unset color there
editor.chain().setTextSelection({ from: 17, to: 22 }).unsetColor().run()

console.log(editor.getHTML())
// expected: alpha keeps its red span (selection never touched it)
// actual:   <ul><li><p>alpha beta</p></li><li><p>gamma delta</p></li></ul>
```

The mark removed isn't even in the selected list item: `nodesBetween` yields the
`bulletList` ancestor (pos 0, `nodeSize` = the whole list), it carries no marks,
so the "no styled textStyle mark" branch runs `removeMark(0, nodeSize)` over the
entire list. The same content in a flat `<p>` behaves correctly, because there
the only visited nodes are the paragraph (skipped as a textblock) and the text
nodes (checked correctly) — which is why this hides outside lists/tables.

**2. Confirmation of the `!node.isText` fix.** We shipped the guard
@nicwegner suggested as an override in our editor and it resolves both the list
and table cases, while an attribute-less leftover span is still cleaned up
within the selection:

```js
removeEmptyTextStyle: () => ({ tr }) => {
    const { selection } = tr
    tr.doc.nodesBetween(selection.from, selection.to, (node, pos) => {
        if (!node.isText) {
            return true // descend, but never removeMark over a container's range
        }
        const marks = node.marks.filter(mark => mark.type === this.type)
        if (marks.length && !marks.some(mark => Object.values(mark.attrs).some(value => !!value))) {
            tr.removeMark(pos, pos + node.nodeSize, this.type)
        }
        return false
    })
    return true
},
```

Only text nodes can carry the `textStyle` mark, so restricting removal to text
nodes is strictly safer than enumerating container types
(`node.type.name.startsWith('table')` etc.). Happy to turn this into a PR
against `main` (+ backport to v2) if maintainers are open to it.
